src/v/cluster/types.h

```diff
@@ -1104,53 +1104,138 @@ struct join_node_reply
   : serde::
       envelope<join_node_reply, serde::version<1>, serde::compat_version<0>> {
     using rpc_adl_exempt = std::true_type;
+
+    enum class status_code : uint8_t {
+        success = 0,
+        // Non-specific error
+        error = 1,
+        // Newly-founded cluster is not yet ready to accept node joins
+        not_ready = 2,
+        // Cluster is busy adding another node, cannot add more until that is
+        // done
+        busy = 3,
+        // Rejected because cluster version violates the bounds in the request
+        incompatible = 4,
+        // Rejected because node's address conflicted with an existing node
+        conflict = 5,
+        // Rejected because node claimed a node_id that differs from the
+        // id it previously registered.
+        id_changed = 6,
+        // Rejected because a node is trying to claim the same node_id with
+        // the same UUID as a node previously removed
+        bad_rejoin = 7,
+    };
+
     bool success{false};
     model::node_id id{model::unassigned_node_id};
 
     std::optional<iobuf> controller_snapshot;
 
+    // Optional because old Redpandas just set the success boolean.
+    // See status() for reading the status, do not read this directly.
+    std::optional<status_code> raw_status;
+
+    status_code status() const {
+        return raw_status.value_or(
+          success ? status_code::success : status_code::error);
+    }
+
+    const char* status_msg() const {
+        switch (status()) {
+        case status_code::success:
+            return "success";
+        case status_code::error:
+            return "error";
+        case status_code::not_ready:
+            return "not_ready";
+        case status_code::busy:
+            return "busy";
+        case status_code::incompatible:
+            return "incompatible";
```

Comment:
can we make those string representations more descriptive they can explain what the error is in human readable form

Reply:
done